### PR TITLE
(fix) O3-2720: Visit stop date input not working in edit mode

### PR DESCRIPTION
**Pull request by vasharma05**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] My work includes tests or is validated by existing tests.

## Summary
This PR removes passing `minDate` and `maxDate` in the visit stop date and visit start date respectively, since the date fields have multiple conditions related to it. These conditions are present when submitting the form here.

Alongside, we display an Inline Notification displaying the error messages.

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-chart/assets/51502471/7e138b74-e880-4cd4-9da9-6e887075cec1

## Related Issue
https://issues.openmrs.org/browse/O3-2720

## Other
None

### CODE REVIEW

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-date-time.component.tsx [hunk 1]
```diff
@@ -107,6 +113,17 @@ const VisitDateTimeField: React.FC<VisitDateTimeFieldProps> = ({
           />
         </ResponsiveWrapper>
       </div>
+      {errors[dateFieldName]?.message && (
+        <InlineNotification
+          aria-label="Close notification"
+          kind="error"
+          onClose={() => setError(dateFieldName, null)}
```
Comment (usamaidrsk) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-date-time.component.tsx:120]: @vasharma05 I think closing the notification, should not remove the error on the field, given that the invalid value will still be existing on the form.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 2]
```diff
@@ -114,37 +115,124 @@ const StartVisitForm: React.FC<StartVisitFormProps> = ({
...
-  }, [t, config, displayVisitStopDateTimeFields]);
+    let maxVisitStartDatetime: Date = null,
+      minVisitStopDatetime: Date = null;
+
+    if (visitToEdit?.encounters?.length) {
+      const allEncountersDateTime = visitToEdit?.encounters?.map(({ encounterDatetime }) =>
+        Date.parse(encounterDatetime),
+      );
+      maxVisitStartDatetime = new Date(Math.min(...allEncountersDateTime));
+      minVisitStopDatetime = new Date(Math.max(...allEncountersDateTime));
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:126]: Spread operators (`...`) seem like a mistake here. `Math.min()` and `Math.max()` do not alter the array, so there's no need to copy them here.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 3]
```diff
@@ -114,37 +115,124 @@ const StartVisitForm: React.FC<StartVisitFormProps> = ({
...
+      ({ date, time, timeFormat }) => {
+        if (!minVisitStopDatetime || !date || !time || !timeFormat) {
+          return true;
+        }
+
+        const [hours, minutes] = convertTime12to24(time, timeFormat);
+
+        const visitStopDatetime = dayjs(date).hour(hours).minute(minutes);
+
+        return visitStopDatetime.isAfter(minVisitStopDatetime);
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:172]: Probably want to clean-up the blank lines.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-date-time.component.tsx [hunk 4]
```diff
@@ -2,26 +2,30 @@ import React from 'react';
...
 import { type amPm } from '@openmrs/esm-patient-common-lib';
 
 interface VisitDateTimeFieldProps {
   visitDatetimeLabel: string;
-  dateFieldName: 'visitStartDate' | 'visitStopDate';
-  timeFieldName: 'visitStartTime' | 'visitStopTime';
-  timeFormatFieldName: 'visitStartTimeFormat' | 'visitStopTimeFormat';
+  datetimeFieldName: 'visitStartDatetime' | 'visitStopDatetime';
   minDate?: number;
   maxDate?: number;
```
Comment (ibacher) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-date-time.component.tsx:23]: Do we still need these?

**Review by ibacher — COMMENTED** [review]
Couple of small things.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx (hunk 2 above)
Comment (vasharma05) [reply packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:126]: Thanks for the catch!

Comment (vasharma05) [reply packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:126]: image
It is the requirement of the `Math.min/max`

**Comment by vasharma05** [comment]
@ibacher @denniskigen requesting review here.
Thanks!

**Comment by rbuisson** [comment]
@vasharma05 , is the screenshot in the PR description up to date?
If so, I don't think it aligns with the feedback provided on Slack [here](https://openmrs.slack.com/archives/CKS32D55G/p1707803262947229).

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-date-time.component.tsx (hunk 1 above)
Comment (usamaidrsk) [reply packages/esm-patient-chart-app/src/visit/visit-form/visit-date-time.component.tsx]: @vasharma05 have you checked this out.
Thanks

**Comment by brandones** [comment]
Hi @vasharma05 , are you just waiting for review? I notice you never replied to @rbuisson 's [comment on Slack](https://openmrs.slack.com/archives/CKS32D55G/p1707919307569549?thread_ts=1707803262.947229&cid=CKS32D55G), nor the one here, and it looks like the screenshot still does not reflect that discussion on Slack. What's going on with this PR? Would be great to wrap this up.

**Comment by vasharma05** [comment]
Hi @brandones!
Yes, I was waiting for a review on the PR, and also, I missed out on replying to Romain and updating screenshots.

I have resolved the file conflicts and updated the GIF in the PR description.
Please review the PR.
Thanks!

**Comment by brandones** [comment]
This appears to collide with https://github.com/openmrs/openmrs-esm-patient-chart/pull/1657 . I suspect you and @usamaidrsk will want to meet/pair and get these two pieces of work done in coordination with each other.

**Comment by brandones** [comment]
There are now conflicts in `visit-date-time.component.tsx` that need to be resolved.

**Comment by denniskigen** [comment]
I suspect this may have already been fixed in https://github.com/openmrs/openmrs-esm-patient-chart/pull/2019. Could you review and confirm, @vasharma05?

**Comment by vasharma05** [comment]
Hi, @denniskigen!
This PR brings some additional changes related to zod validation, and correct validations among start and end visit date and time. 
I'll close this PR for now, as it's lingering for some time now, and will make changes on the newest changes done till date.
Thanks for all the review!

**Pull request closed without merging**